ert3/data/_record.py

```diff
@@ -91,3 +105,188 @@ def ensure_consistent_record_names(cls, multi_ensemble_record):
 
     def __len__(self):
         return len(self.record_names)
+
+
+class RecordTransmitterState(Enum):
+    transmitted = auto()
+    not_transmitted = auto()
+
+
+class RecordTransmitterType(Enum):
+    in_memory = auto()
+    ert_storage = auto()
+    shared_disk = auto()
+
+
+class RecordTransmitter:
+    def __init__(self, type_: RecordTransmitterType):
+        self._type = type_
+        self._state = RecordTransmitterState.not_transmitted
+
+    def _set_transmitted(self):
+        self._state = RecordTransmitterState.transmitted
+
+    def is_transmitted(self):
+        return self._state == RecordTransmitterState.transmitted
+
+    @abstractmethod
+    async def dump(self, location: Path) -> None:
+        pass
+
+    @abstractmethod
+    async def load(self) -> Record:
+        pass
+
+    @abstractmethod
+    async def transmit_data(
+        self,
+        data: Union[List[float], Mapping[int, float], Mapping[str, float], List[bytes]],
+        mime,
+    ) -> None:
+        pass
+
+    @abstractmethod
+    async def transmit_file(
+        self,
+        file: Path,
+        mime,
+    ) -> None:
+        pass
+
+
+class SharedDiskRecordTransmitter(RecordTransmitter):
+    TYPE: RecordTransmitterType = RecordTransmitterType.shared_disk
+
+    def __init__(self, name: str, storage_path: Path):
+        super().__init__(type_=self.TYPE)
+        self._storage_path = storage_path
+        self._storage_path.mkdir(parents=True, exist_ok=True)
+        self._concrete_key = f"{name}_{uuid.uuid4()}"
+        self._uri: typing.Optional[str] = None
+        self._mime = ""
+
+    def set_transmitted(self, uri: Path, mime: str):
+        super()._set_transmitted()
+        self._uri = str(uri)
+        self._mime = mime
+
+    async def _transmit(self, record: Record, mime: str):
+        storage_uri = self._storage_path / self._concrete_key
+        if mime == "application/json":
+            contents = json.dumps(record.data)
+            async with aiofiles.open(storage_uri, mode="w") as f:
+                await f.write(contents)
+        else:
+            if isinstance(record.data, list) and isinstance(record.data[0], bytes):
+                async with aiofiles.open(storage_uri, mode="wb") as f:  # type: ignore
+                    await f.write(record.data[0])  # type: ignore
+            else:
+                raise TypeError(f"unexpected record data type {type(record.data)}")
+        self.set_transmitted(storage_uri, mime)
+
+    async def transmit_data(
+        self,
+        data: Union[List[float], Mapping[int, float], Mapping[str, float], List[bytes]],
+        mime,
+    ) -> None:
+        if self.is_transmitted():
+            raise RuntimeError("Record already transmitted")
```

Comment:
This seems like guarding unexpected usage of the transmitter. Would an `AssertionError` be more fitting? 🤔

Reply:
Assertion in production? 😬

Reply:
I retract my comment..